### ISSUE TITLE
`Box<T, A: AllocFrom>`

### ISSUE THREAD

**Issue by scottjmaddox** [rust-lang/wg-allocators#12]
The current push to associate an allocator to boxes (among other types) is targeting something like `Box<T, A: Alloc>`. This limits the design space of custom allocators, though, because it forces one of two approaches for accessing the `Alloc`:

1. Use a single static global state (or pointer or reference to the state).
2. Embed a pointer or reference to the state inside every box.

There are many kinds of allocators that these limitations rule out. For example, perhaps you want to have multiple independent "Arena" or "Region" allocators that each allocate from a dedicated slab of memory. Since you want to have multiple independent allocators, you cannot (efficiently) use approach 1. So you're forced to use approach 2, which is extremely inefficient, requiring a whole extra pointer to be stored in each `Box`.

One approach for avoiding this limitation would be something like the following:

```rust
pub struct Box<T: ?Sized, A: AllocFrom>(Unique<T>, A);

pub trait AllocFrom {
	type Alloc: Alloc;

    fn alloc_from<T: ?Sized>(&self, heap_ptr: *const T) -> Self::Alloc;
}
```

Note that `AllocFrom` provides the method `alloc_from`, which accepts `&self` and a pointer to heap memory and returns an `Alloc` implementation. This gives many options for locating the allocator's state and generating an `Alloc`. Approach 1 (global state) is of course still possible, and so is approach 2 (embed a pointer in every box). But this also makes approach 3 possible:

3. Calculate a pointer to the allocator state from a pointer to heap memory allocated by that allocator.

This makes many more clever and efficient kinds of allocators possible, such as an allocator that allocates all 8-byte allocations from aligned 1 MiB blocks, and stashes a pointer to its own state at the beginning of each of those 1 MiB blocks. Such an allocator can implement `AllocFrom` to check the size of the allocated memory using `mem::size_of<T>()` and if 8-bytes then mask off the lowest 10 bits from `heap_ptr` to get a pointer to the allocator's state, from which it can construct an `Alloc`.

To clarify, `AllocFrom` makes it possible to calculate a pointer to the allocator's state from the `Box` itself:

```rust
impl<T: ?Sized, A: AllocFrom>  Box<T, A> {
    fn get_alloc(&self) -> A::Alloc {
        self.1.alloc_from(self.0.as_ptr())
    }
}
```

Since the `Box` provides a pointer to memory allocated by the allocator, the allocator can use its control of the pointers it returns and the memory around those pointers to efficiently locate its own state and generate an `Alloc`. This makes many more kinds of clever and efficient allocators possible.

Relevant comments from [PR #58457: Associate an allocator to boxes](https://github.com/rust-lang/rust/pull/58457):
- https://github.com/rust-lang/rust/pull/58457#issuecomment-466681204
- https://github.com/rust-lang/rust/pull/58457#issuecomment-466739043
- https://github.com/rust-lang/rust/pull/58457#issuecomment-475089881
- https://github.com/rust-lang/rust/pull/58457#issuecomment-482816662
  - this comment has the most complete usage example, although it defines `AllocFrom` to have a method for each container type, when really we just need a single method that accepts a pointer to heap memory

**Comment by TimDiekmann** [comment]
So if I understand correctly, `AllocFrom` is related to `Alloc` just like `BuildHasher` is related to `Hasher`?

**Comment by scottjmaddox** [comment]
> So if I understand correctly, `AllocFrom` is related to `Alloc` just like `BuildHasher` is related to `Hasher`?

Almost exactly. The method signature is different, though, because the whole point is to give the allocator access to a pointer to its own heap memory from which to generate its `Alloc` implementation.

**Comment by glandium** [comment]
I do have this very use-case, actually. And it didn't strike me that it could be solved this way, and I must say I like it. But there are details to figure out. Like... how do you get the `AllocFrom` value you attach to the `Box` from the `Alloc.alloc`? Because `Box::new_in` would have to take an `Alloc` as input, and the `AllocFrom` would need to be created somehow.

(Note that last iteration in https://github.com/rust-lang/rust/pull/58457 doesn't have bounds on the `Box` type itself at all. They're all at the implementation level. Which actually gives some additional flexibility)

**Comment by SimonSapin** [comment]
Aren’t bounds on the type definition required to allow `impl Drop` to have the same bounds?

**Comment by TimDiekmann** [comment]
> Calling it `BuildAlloc` for consistency would be perfectly reasonable, though.

This is exactly what I had in mind. Do you want to change the issue and the title to use `BuildAlloc` instead of `AllocFrom`?

**Comment by scottjmaddox** [comment]
> how do you get the `AllocFrom` value you attach to the `Box` from the `Alloc.alloc`? Because `Box::new_in` would have to take an `Alloc` as input, and the `AllocFrom` would need to be created somehow.

Yes, the type signature for `new_in` would need to be something like this:

```rust
impl<T: ?Sized, A: BuildAlloc>  Box<T, A> {
    fn new_in(&self, alloc: A::Alloc) -> Box<T, A>;
}
```

And, as you point out, you need some way to create a `BuildAlloc` from an `Alloc`. So a new method on `Alloc` will need to be added, something like `get_build_alloc`, and an associated type, `type BuildAlloc: BuildAlloc`.

I wasn't sure if such cyclic associated types are allowed, [but it looks like they are](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=89751a6f25aed206a9e3ae1bf1c087ac).

**Comment by glandium** [comment]
> Aren’t bounds on the type definition required to allow `impl Drop` to have the same bounds?

There's an interesting problem there... Drop currently does nothing. It's actually implemented by the compiler, which then goes on generating calls to `box_free` on its own.

**Comment by glandium** [comment]
> So a new method on Alloc will need to be added, something like get_build_alloc

Are there use cases where different allocations could have a different `BuildAlloc` that would point back to the same `Alloc`? Those would be left out with this approach.

**Comment by scottjmaddox** [comment]
> Are there use cases where different allocations could have a different `BuildAlloc` that would point back to the same `Alloc`? Those would be left out with this approach.

None that I can think of. That doesn't mean they don't exist, though. Allocators traditionally use a single global `free` function that accepts a pointer to the memory to be freed and calculates a pointer to the allocator from that (or from global static state). So all of the history of allocator development has assumed the equivalent of a single `BuildAlloc` per allocator.

Ultimately, you can put whatever logic you want in `BuildAlloc`. So it's more a question of can what you want to do be done more efficiently with multiple `BuildAlloc`s per `Alloc`. And the answer is probably "maybe".  But if that's the only way to support `BuildAlloc` with the current type system, then it's still a huge improvement over the limitations inherent to `Box<T, A: Alloc>`. Don't let perfect be the enemy of good, and whatnot.

**Comment by TimDiekmann** [comment]
How could `BuildAlloc` play together with #9?

**Comment by gnzlbg** [comment]
>  Since you want to have multiple independent allocators, you cannot (efficiently) use approach 1.

I'm not sure why this is the case. If you put whatever you need in a `static`, or in a thread local, you can just have a `ZST` as an allocator, whose methods access it.

**Comment by scott-maddox** [comment]
> > Since you want to have multiple independent allocators, you cannot (efficiently) use approach 1.
> 
> I'm not sure why this is the case. If you put whatever you need in a `static`, or in a thread local, you can just have a `ZST` as an allocator, whose methods access it.

But they're not really multiple independent allocators then, are they? They're one allocator that pretends to be multiple allocators. And there's an overhead associated with that. If the allocators are truly separate, then you could, for example, safely make them `!Send` and `!Sync` and do away with all thread synchronization overhead (and avoid thread local storage overhead).

**Comment by gnzlbg** [comment]
> But they're not really multiple independent allocators then, are they?

This is what I meant:

```rust
struct A; // my allocator type
fn foo() {
    static A0: A = A; // allocator 0
    static A1: A = A; // allocator 1
    struct AH0; // ZST handle to 0
    impl Alloc for AH0 {
        fn alloc(...) -> ... { A0.alloc(...) }
    }
    // ... same for AH1 ...
    let b0: Box<T, AH0>; // box using A0
    let b0: Box<T, AH0>; // another box using A0
    let b1: Box<T, AH1>;  // box using A1
    // ^^ all boxes are pointer sized cause the handle is a ZST
}
```

Whether you can have multiple handles to the same allocator or not, depends on your allocator impl. Whether you want to have one or multiple allocators, with different handles, is kind of up to you. 

If you can put your allocators in statics, or thread locals, then you can have zero-sized handles to them. If you cannot, then your handle needs to have a pointer to the allocator, but I don't think there is any way around that.

**Comment by SimonSapin** [comment]
@gnzlbg This uses different Rust types for different handles, so there’s a bounded number of such types determined at compile-time.

This is not what people mean by “multiple instances” of an allocator. For example:

```rust
fn library_entry_point(some_input: Foo) {
    let handle = ArenaAllocator::new();
    // …
    let b = Box::new_in(something, handle.clone());
    let v = Vec::new_in(handle.clone());
    // …
}
```

A program that uses this library could spawn a large number of threads that each have their own separate arena. At compile time you don’t how many threads there’s gonna be, so you can have a Rust type for each of them.

For `v` to be able to tell which of these arenas it belongs to, `handle` cannot be zero-size. Or, conversely, if the handle is zero-size then the allocator must have global state that is shared with all handles of the same type.

**Comment by gnzlbg** [comment]
@SimonSapin makes sense, I was misunderstanding which problem this solves, `BuildAlloc` feels useful.

---

All `Alloc` methods that need to identify an allocation take a pointer to an allocation (e.g. a `NonNull<u8>`) and its `Layout`. Why is `BuildAlloc::build_alloc`:

```rust
fn build_alloc<T: ?Sized>(&self, heap_ptr: *const T) -> Self::Alloc;
// nitpick: heap_ptr does not necessarily point to the heap
```

generic over `T` instead of 

```rust
fn build_alloc(&self, ptr: *const u8, layout: Layout) -> Self::Alloc;
```

?

**Comment by scott-maddox** [comment]
@gnzlbg You're right, it should be `unsafe fn build_alloc(&self, ptr: NonNull<u8>, layout: Layout) -> Self::Alloc;`, in order to match `Dealloc`. I'll update the OP.

The reason I had proposed making the API safe is because the implementer must agree to ensure it is safe. For example, the implementer **must not** de-reference the heap_ptr (unless it somehow knows it can do so safely). However, we need to pass the `Layout` separately to support all possible allocations, and thus the interface has the same potential for misuse as `Dealloc` and thus must be marked `unsafe`.

**Comment by scottjmaddox** [comment]
@gnzlbg You're right, it should be `unsafe fn build_alloc(&self, ptr: NonNull<u8>, layout: Layout) -> Self::Alloc;`, in order to match `Dealloc`. I'll update the OP.

The reason I had proposed making the API safe is because the implementer must agree to ensure it is safe. For example, the implementer **must not** de-reference the heap_ptr (unless it somehow knows it can do so safely). However, we need to pass the `Layout` separately to support all possible allocations, and thus the interface has the same potential for misuse as `Dealloc` and thus must be marked `unsafe`.

**Comment by scottjmaddox** [comment]
I also can't think of any reason `build_alloc` shouldn't take `&mut self`. You are guaranteed to have a `&mut self` during dealloc and realloc. If you are cloning a `&Box`, you are guaranteed to be able to clone the `BuildAlloc` which you could then get a `&mut self` to.

There might be some use cases where this could come in handy; maybe tracking the number of reallocations within the `BuildAlloc`?

So I'll change that as well.

**Comment by TimDiekmann** [comment]
@scottjmaddox You may take a look at https://github.com/TimDiekmann/alloc-wg? I have used `BuildAlloc` as parameter for `Box` and `RawVec`. Is this the way you had in mind?
ZST allocators like `Global` and `System` implements both, the builder and the allocator.

**Comment by Ekleog** [comment]
I've just come across this issue. As it suggests adding a new function that makes it possible to have better control over how exactly the allocation is performed, would it be possible to take advantage of this method addition to make the method fallible, so that it's also possible to control how to behave when the allocation fails? (One example use case would be using an arena allocator for some untrusted data coming from the network, and wanting to abort just this specific client's instance if something OOMs here)

**Comment by SimonSapin** [comment]
Fallibility is a separate "axis" from using a provided allocator v.s. the global one. All four combinations are potentially useful.

**Comment by Ekleog** [comment]
They are, but only the version that takes both an allocator and is fallible is required to be implement all the other ones: it's easy to do `Box::new_in(t, alloc).unwrap_oom()` or `Box::new_in(t, std::alloc::GLOBAL_ALLOCATOR)` (assuming here a function like `unwrap_oom` that'd do the proper `handle_alloc_error` on `Result<T, AllocErr>`, and a global variable that allows read-only access to the global allocator).

Now, if you want to make all four versions it's great too :) what I want to point out is just that if you plan on adding a single function for the time being, then it'd probably be better if it were the fallible one. (Also, I'm not sure we want to encourage the whole allocating ecosystem to provide 4 functions for each use case by setting that example in libstd)

**Comment by SimonSapin** [comment]
I should have mentioned the RFC https://rust-lang.github.io/rfcs/2116-alloc-me-maybe.html and its tracking issue https://github.com/rust-lang/rust/issues/48043 for `try_reserve` on collections both have discussion of what the APIs should look like, for both this minimal set of methods and a future, more comprehensive fallible methods (`try_push`, `try_extend`, …)

Yes `Box::try_new_in` is the most basic/powerful combination that others can be defined on top of, and trying to reduce multiplicative explosion of the number of methods is a good point, but exposing it still involves making API design decisions.

**Comment by TimDiekmann** [comment]
I like to resolve the issues which proposes to change `AllocRef` before starting to implement it for collections. I have played around with `BuildAllocRef` a lot a while ago and I liked the concept in general. However, for most use cases, it's probably overkill. On the other hand, the common Rust user won't implement many allocators theirself, so this shouldn't be a that big problem.

IMO in order to go further with this proposal, we have to fledge out an interface, where each signature should be fully defined. Currently, it's a broad idea, of where to build `AllocRef` and where to pass `BuildAllocRef`. For testing purposes we either could use a branch on the `alloc-wg` crate or I try to keep things updated upstream.

I'll throw in some points here:
- We probably need an associative type on both `AllocRef` and `BuildAllocRef`.
- For collections we want the generic parameter to be `A: AllocRef` instead of `B: BuildAllocRef`, as type inference would not work otherwise (see https://github.com/TimDiekmann/alloc-wg/issues/5).
- What do we store inside containers (`BuildAllocRef` vs `AllocRef`)? What to we want to pass to (e.g.) `Vec::new_in`?
  for `BuildAllocRef`: 
  - Is it efficient enough to build `AllocRef` every time we need it?
  for `AllocRef`:
  - How we get a `BuildAllocRef` from `AllocRef`? I think this is needed for cloning for example.

**Comment by TimDiekmann** [comment]
I have tested around a bit again with `BuildAllocRef` using more or less exactly the code from the OP. You can see the code in the [buildalloc-test branch at alloc-wg](https://github.com/TimDiekmann/alloc-wg/tree/buildalloc-test). You have to clone it manually and create a doc to view it, as I haven't uploaded it somewhere yet.

Some notes on the implementation:
- For `alloc`, there is no pointer. Even if you have a `Layout` available (the one you want to use for `alloc`), it's feeling unintuitive. The idea was, that we pass `NonNull::dangling()`, but a new method for creating an `AllocRef` may be useful. Maybe `build_alloc_ref` and `get_alloc_ref`, or `build_alloc_ref` and `create_alloc_ref`, or `new_alloc_ref` and `build_alloc_ref`?
- I haven't tested around implementing `BuildAllocRef` for other types than ZSTs yet.
- Every test passed without any changes

**Comment by Amanieu** [comment]
One thing that I feel hasn't been fully addressed is how you can construct a `Box<T, BuildAllocRef>` where the `BuildAllocRef` is a ZST but the `AllocRef` isn't. This seems important as this is the main use case for having `BuildAllocRef` at all.
